bibblio.py

```diff
@@ -168,6 +167,116 @@ def create_content_item(self, content_item):
+            disallowed_response_codes=['4xx']
+        )
+
+        if not isinstance(identifier, basestring) and response.status_code == 200:
+            self._db.delete(identifier)
+            self.log.info("DELETED: Bibblio Content Item '%s'" % content_item_id)
+
+
+class EpubFilter(object):
+
+    """A base class for source-specific EPUB filtering. This class
+    removes front matter and distributor-specific text that can impact
+    recommendations created by the BibblioAPI.
+    """
+
+    FILLER_RE = '\s*'
+    PUNCTUATION_RE = '(\.|,|-|;)*'
+
+    ## Values for subclass definition ##
+
+    # SPINE_IDREFS lists idref values in the EPUB spine that can be
+    # completely ignored, usually because they're chock full of text
+    # specific to the distributor without any useful text to support
+    # recommendations.
+    SPINE_IDREFS = None
+
+    # FILTERED_PHRASES lists strings that are directly related to the
+    # distributor in decreasing order of specificity.
+    FILTERED_PHRASES = None
+
+    @classmethod
+    def filter_spine_idrefs(cls, spine_idrefs):
+        for s in spine_idrefs:
+            if s in cls.SPINE_IDREFS:
+                print "'%s' in SPINE_IDREFS" % s
+        return [s for s in spine_idrefs if s not in cls.SPINE_IDREFS]
+
+    @classmethod
+    def phrase_regex(cls, phrase):
+        """Incorporates whitespace catchall string into a phrase"""
+        words = [word for word in phrase.split() if word]
+        phrase = cls.FILLER_RE.join(words)
+        return re.compile(phrase, re.IGNORECASE)
+
+    @classmethod
+    def filter(cls, text):
+        filtered_text = text
+
+        for phrase in cls.FILTERED_PHRASES:
+            phrase_re = cls.phrase_regex(phrase)
+            filtered_text = re.sub(phrase_re, '', filtered_text)
+        return filtered_text
+
+
+class GutenbergEpubFilter(EpubFilter):
+
+    SPINE_IDREFS = set(['pg-header'])
+
+    FILTERED_PHRASES = [
+        (
+            'This eBook is for the use of anyone anywhere (in the United'
+            ' States)? (and)? (most other parts of the world)? at no'
+            ' cost and with almost no restrictions whatsoever.'
+        ),
+        (
+            'You may copy it, give it away or re-use it under the terms'
+            ' of the Project Gutenberg License included with this eBook'
+            ' or online at'
+        ),
+        (
+            'If you are not located in the United States, you\'ll have'
+            ' to check the laws of the country where you are located'
+            ' before using this ebook.'
+        ),
+        '(http)?s?(:\/\/)?www\.gutenberg\.(org|net)(\/)?(\w|\.|-)*',
+        '(The)? Project Gutenberg Ebook (of)?',
+        '(The)? Project Gutenberg E(-)?text (of)?',
+        'Project Gutenberg License',
+        'Project Gutenberg',
+        'Gutenberg',
+    ]
+
+    # DISQUALIFYING_PHRASES = [
+    #     'START : FULL (_)? LICENSE (\.|\*|\n)* THE FULL PROJECT GUTENBERG LICENSE'
+    # ]
```

Comment:
Ugh, this is not supposed to be in the commit. Some Gutenberg books have a file that just has license information in it and nothing else. It doesn't have a consistent name, but it has some consistent text.

Unfortunately, some of them just shove the license into an existing file that is the only file and has the whole book in it. So this was sending empty files up to API and got cut.